### DISCUSSION TITLE
ADO and DAO replacement?

### DISCUSSION THREAD

**Discussion by EduardoVB** [twinbasic/twinbasic#1012] [Ideas]
I now need to move [one VB6 project](https://www.vbforums.com/showthread.php?897088-Database-alternative-(to-store-much-data)) from an Access database to something that could handle a larger database.

I was suggested SQLite, that I had already in mind.

I always used DAO. Why? Because I like it. Because it is simple, straightforward, powerful, fast and mainly it has a VBish like interface and way to use it.
And also because I made my first large program in the times when ADO didn't exist and never saw a need to change to anything else (until now).

Yes, there are some lacking features, the only one I ever needed is disconnected recordsets. In those cases I used ADO.

Also making the database design in Access is a piece of cake, so friendly.

But to the point. I now need to move to SQLite, and I was thinking it would be great to have a component with the DAO interface, and also another one with the ADO interface for the ones that use ADO, that are the majority.

Krool, made a [component for using SQLite in VB6](https://www.vbforums.com/showthread.php?882857-VB-SQLite-Library-(COM-Wrapper)). 
I think in top of that, or modifying it, a component could use the DAO or ADO interface.
I don't think I would embrace in such a project for VB6 now, it would be an important task and I have to work in other things.

But maybe it could be done for tB that has a future (by me or someone else). What do you think?
I don't mean a full 1:1 replacement with all the features (all types of supported recordsets, etc), but at least the main and more commonly used ones.
Does it make sense?

PS: of course if I would ever do that I'll have to learn a lot. Maybe someone else is more ready.
For now I have in mind just to change my code to use Krool's component. A couple of days work vs a couple of months work (or more).

**Comment by bclothier** [comment]
I am not so sure ADO needs to be "replaced". It's a part of WDAC and continues to be supported. At one point, Microsoft was going to deprecate SQL Server support for OLEDB (which ADO builds on top of) but then backtracked and has remained committed to supporting OLEDB. 

Both ODBC and OLEDB (as well ADO) are now considered [Windows components and will be shipped with Windows](https://docs.microsoft.com/en-us/previous-versions/windows/desktop/ms692897(v=vs.85)). Also, see [support policy for OLEDB](https://docs.microsoft.com/en-us/sql/connect/oledb/applications/support-policies-for-oledb-driver-for-sql-server?view=sql-server-ver16) and [support for ODBC](https://docs.microsoft.com/en-us/sql/odbc/microsoft/product-support?view=sql-server-ver16). 

In the case of DAO, you can just use Access redistributable engine ([latest would be 2016](https://www.microsoft.com/en-us/download/details.aspx?id=54920) or  if you want to continue using .MDB/.ACCDB files as your data store. This would allow you to continue DAO as-is. In case of the Office365 installation, there's also [Access 365 Runtime](https://support.microsoft.com/en-us/office/download-and-install-microsoft-365-access-runtime-185c5a32-8ba9-491e-ac76-91cbe3ea09c9). [More information about choosing between MSI vs. C2R installs](https://docs.microsoft.com/en-us/office/troubleshoot/access/cannot-use-odbc-or-oledb). 

In the case of sqlite, there is nothing stopping us from creating a DAO driver for SQLite because we have capabilities for implementing the DAO interfaces. YMMV.

**Reply by EduardoVB** [reply]
> In the case of sqlite, there is nothing stopping us from creating a DAO driver for SQLite because we have capabilities for implementing the DAO interfaces. YMMV.

Do you know how to do it?

**Reply by bclothier** [reply]
Actually, anybody who can implement an interface, can too. Unlike VBx, tB is far more flexible with allowing you to implement the interfaces to support a particular API.

**Reply by EduardoVB** [reply]
> anybody who can implement an interface, can too

I can implement an interface but would have no idea where to start to make such a driver, and that actually works well and mimic what Jet 4 does with MDB files ([Seek](https://docs.microsoft.com/en-us/office/client-developer/access/desktop-database-reference/recordset-seek-method-dao), etc).

**Comment by EduardoVB** [comment]
OK, but doing so we keep relying on MS.

**Reply by mburns08109** [reply]
> OK, but doing so we keep relying on MS.

Well for the tB 1.0 world, that's kind-of a design requirement, isn't it?
Come to think on it..."Not being totally dependent on a particular software platform" may well be a good language design goal for the "tB > ver. 1.0" future.

**Reply by EduardoVB** [reply]
OK, as a reference we are continuing a discussion from [here](https://github.com/twinbasic/twinbasic/issues/999).

> Well for the tB 1.0 world, that's kind-of a design requirement, isn't it?

No. The requirement is being backward compatible with VB6, not dependent on MS.
But I was talking about the core language, not about other things or packages.
WinForms and other things will always be dependent on Windows (that's why they were named "WinForms" in tB). In some future I hope you'll be able to choose Windows only project, cross-platform project (limited to what every target platform can support) or other project types I guess (for example Andoid only, or WebAssembly).
That's my understanding.

The core language will be more or less the same for everyone. 
I was taking about the things added to the core language (for not making it Windows dependent), not about everything.
What you asked about the Now function is in the core language.

**Comment by EduardoVB** [comment]
Anyway I didn't mean to make a full DAO/ADO replacement, but just something that can work with SQLite using those interfaces.

ADO and DAO are database engines that can work with many database formats, but AFAIK there is no driver for SQLIte for Jet4.

The idea would be to replace DAO/ADO and MDB with something the like but using SQLite. Not making a full database engine, that would be a tasks of another magnitude.

IDK if it is a good idea or not.

**Reply by mburns08109** [reply]
Well, since there is are some ODBC Drivers for SQLLite available (here's one: https://github.com/softace/sqliteodbc but I saw at least 2 others available too, if proprietary)
I would think that either DAO or ADO would still be reasonable solutions for you.

**Reply by EduardoVB** [reply]
> Well, since there is are some ODBC Drivers for SQLLite available (here's one: https://github.com/softace/sqliteodbc

Thanks. I would have to test it, but it seems it works with earlier SQLite versions and seems abandoned now. Anyway it might work.

**Reply by EduardoVB** [reply]
I could connect with DAO using [that driver](http://www.ch-werner.de/sqliteodbc/), but the [Seek method](https://docs.microsoft.com/en-us/office/client-developer/access/desktop-database-reference/recordset-seek-method-dao) in the table mode recordset, that is what I needed (because it is the fastest, a lot faster than FindFirst) is not available for such connection types unfortunately.

And BTW I'll leave here the way to connect SQLite with DAO via ODBC, since I spent a while figuring it until it worked:

```
    Dim iWs As dao.Workspace
    Dim iDAOSQLiteDb As dao.Database
    Dim iRecDAOSQLite As dao.Recordset
    
    Set iWs = DBEngine.CreateWorkspace("", "admin", "", dbUseODBC)
    Set iDAOSQLiteDb = iWs.OpenDatabase("", , , "ODBC;DRiver=SQLite3 ODBC Driver;Database=" & App.Path & "\MySQLite.db;")
    Set iRecDAOSQLite = iDAOSQLiteDb.OpenRecordset("TableName")

```

Update: unfortunately when trying to use a recodset from a SQL query, created with it, I get random errors. So I ruled out using that driver.

**Reply by VB6-Programming** [reply]
An often simpler way is to create an Access linked table pointing to your back-end database via ODBC.  I've not done this with SQLite but I have often used this with MS SQL Server, SQL Server Express, Oracle and other databases.

The advantage of this approach is that you do not need to modify the VB6 app. You just use the Access linked database (named the same and in the same location as your existing Access database)  and this uses the data from the back-end database.
Also it lets you support multi-user and even move to a cloud-based database.

But, as you pointed out, this won't work with table-type recordsets using the Seek method.

(A separate issue is why you are using Seek.  As you say it is faster than FindFirst.  But you should be letting SQL do the searching for you.  I've found that with properly indexed SQL databases, using a SQL Query is faster than opening an entire table to use Seek.
But this all depends on what you are doing. Often in my typical CRUD business apps I only require to find one record in a table, your requirements may be different.
It will also vary depending on the back-end database, you may get different performance with a client/server database such as SQL Server than you would with SQLite.).

**Reply by bclothier** [reply]
A linked table uses ODBC internally so that only adds more layers but won't help if the SQLite ODBC driver is flaky. My experience has been that it's poorly maintained & implemented, which is why I suggest native APIs. The native APIs can be then wrapped in DAO interfaces to provide familiarity for those accustomed to working with DAO but the capabilities ultimately depend on what the native API offers.

**Reply by VB6-Programming** [reply]
Yes, I've never used any SQLite ODBC driver (there are others including https://www.devart.com/odbc/sqlite/download.html )
Though the driver EduardoVB used seems to be highly regarded elsewhere. Here is an outline to installing it for an Access linked table https://how-to.aimms.com/Articles/118/118-Connect-SQLite.html  that may be useful.

The advantage of the linked table approach is simplicity, you can go from a single-user Access database to multi-user Access, to client/server databases (SQL Server, Oracle etc.) to cloud (Microsoft Azure SQL) all without requiring any modification to your VB6 app.
But it does, of course,  require you to be able to connect to the back-end database and you can't use the Seek method (not table-type recordsets).

It would be interesting to try connecting using just Microsoft Access to SQLite.

**Reply by EduardoVB** [reply]
> the driver EduardoVB used seems to be highly regarded elsewhere.

Maybe with other languages, IDK. In my testing it worked fine if I went in step by step mode (F8) but as soon as I pressed F5 got an error (with a big number) when trying to read a field of the recordset.

> (A separate issue is why you are using Seek. As you say it is faster than FindFirst. But you should be letting SQL do the searching for you. I've found that with properly indexed SQL databases, using a SQL Query is faster than opening an entire table to use Seek.
> But this all depends on what you are doing.

It depends very much and what we are doing, yes. For many applications speed won't matter so much, I know.

Seek is really fast, quite faster than making new SQL queries each time too.
I guess most people don't know how it performs and that's why it is not so popular (and also because it is supported just in Access databases). Many years ago I didn't know either and used FindFirst everywhere. But Seek compared to FindFirst is... I don't remember the exact numbers now but like 1000 times faster.

But guys, thank you but I think I have [my current issue resolved](https://www.vbforums.com/showthread.php?882857-VB-SQLite-Library-(COM-Wrapper)&p=5574823&viewfull=1#post5574823). (Well, I'm still converting the code, but it seems it will work)

**Reply by EduardoVB** [reply]
But to the point here: my idea was the possibility to be able to use the existent codebase that is using DAO and ADO as it is, or with minimal changes and move to a new database system that could be SQLite (or another one) without the Access limits and that could be in the future easily migrated to other OSs.
That was the intention. It was not about how to solve my current need (that BTW is in VB6). Thank you.

**Reply by mburns08109** [reply]
EduardoVB,

just out of curiosity, why did you go with SQLLite as an Access/DAO alternative?
I mean, if it was an externally-imposed requirement, cool, but if not, mySQL or even SQL Server LocalDB FILE-storage (up to 10GB) are equally if not better/more-proven alternatives. Then there's several web-based options...?
(not a criticism, just curious as to why you made that selection)

**Reply by EduardoVB** [reply]
Because [it was recommended to me](https://www.vbforums.com/showthread.php?897088-RESOLVED-Database-alternative-(to-store-much-data)) and since it is free, could work in other OSs in the future, has no practical limit (281 Terabytes), and also because it was the only suggested option (and I don't know much about the alternatives).
You are invited to join that discussion, but know it: Olaf is tough 😄.

**Reply by jpbro** [reply]
SQLite is great, and [well "proven"](https://www.sqlite.org/mostdeployed.html). One of the big advantages for many applications is that it doesn't require you/your users to install a database server like MySQL/SQL Server since it's a file-based DB.

**Reply by VB6-Programming** [reply]
SQLite is very solid and very, very widely used. It takes a lot of beating.  Though in many business applications a true multi-user, client/server database like SQL Server (or the free SQL Server Express) or Oracle may be preferred.

**Comment by rexxitall** [comment]
ODBC is a snail. Just to avoid the BS word :) Here is something is use for years https://github.com/govert/SQLiteForExcel/tree/master/Source. Works great. MYSQl and friends have the disadvantage that they use a lot of files for one database. And its hard to migrate if they change something. Last months i was using SQLITE to get 80 Millions of drone 3D laser scan RGB data organised. Was maybe not on maximum speed like some advanced DB systems would have done. But was implemented within less a week and worked. Civil engineers does not have the time like a developer. There is always a cruel deadline to be ready ). 9 bridges 2 months. The fun with the data we saw first as we got the ASCII files of them. After we forced them to send them by USB stick and postal mail. I was inside a time travel back to the 80ties. They have had them stored in some kind of cloud which was not able to transfer that "masses". I love Germany ... Anyhow SQLITE is at all not a bad choice. I use it now since i guess 7 years in VBA and i do not miss access anymore. The beauty is that its all in one file and you can use simple SQL statements. No funny tricks. The API is open, bindings are available for "whatever" which makes it very  flexible. And best its free of charge :)
And yes Access and SQL was my first touch with VBA. They forced me to write a database for "unsharp" text search like google it does now (Final engineering diploma thesis). Stupidly i did'nt know VBA, Access,Excel and SQL. At that days i was on a Powerbasic trip. Most stupidly the company dies after i was finished with it. And i am *pi* til today cause i was to stupid to understand what i have done there. Works quite well. You could put ASCII files from price offers in it and you could search in it via some kind of full text research (WITH  DUI !!!). Forget GOOGLE !!! *laugh* They would never exists if... If it looks funny to you. It was around 1990 and internet was done by a modem at that days. Comps has a max speed of 200 MHZ and then you was a lucky owner of a 32 bit monster. Anyhow i got my Diploma with a "good" which was not bad at that school. I guess you has have to do something like Einstein to get a very good. If you reach this, yes it compiles already in twinbasic ;)
